hare/broker.go

```diff
@@ -156,34 +169,72 @@ func (b *Broker) eventLoop() {
 	}
 }
 
+func (b *Broker) updateCurrentLayer(id InstanceId) {
+	if id > b.currLayer { // update only newer layers
+		b.currLayer = id
+	}
+}
+
+func (b *Broker) updateSynchronicity(id InstanceId) {
+	if _, ok := b.outbox[id]; ok { // already has result
+		return
+	}
+
+	// not exist means unknown, check & set
+
+	if !b.isNodeSynced() {
+		b.outbox[id] = nil // mark invalid
+		return
+	}
+
+	b.outbox[id] = make(chan *Msg, InboxCapacity) // mark valid
+}
+
+func (b *Broker) isSynced(id InstanceId) bool {
+	b.updateSynchronicity(id)
+
+	ch, ok := b.outbox[id]
+	if !ok { // not exist means unknown
+		log.Panic("outbox doesn't contain a value after call to updateSynchronicity")
+	}
+
+	if ch == nil { // if exist & invalid
+		return false
+	}
+
+	return true // exist & valid
+}
+
 // Register a listener to Messages
 // Note: the registering instance is assumed to be started and accepting Messages
-func (b *Broker) Register(id InstanceId) chan *Msg {
-	inbox := make(chan *Msg, InboxCapacity)
-
-	wg := sync.WaitGroup{}
-	wg.Add(1)
+func (b *Broker) Register(id InstanceId) (chan *Msg, error) {
+	res := make(chan chan *Msg, 1)
 	regRequest := func() {
-		if id > b.maxReg {
-			b.maxReg = id
-		}
+		b.updateCurrentLayer(id)
 
-		b.outbox[id] = inbox
-
-		pendingForInstance := b.pending[id]
-		if pendingForInstance != nil {
-			for _, mOut := range pendingForInstance {
-				b.outbox[id] <- mOut
+		if b.isSynced(id) {
+			pendingForInstance := b.pending[id]
+			if pendingForInstance != nil {
+				for _, mOut := range pendingForInstance {
+					b.outbox[id] <- mOut
```

Comment:
there is a potential deadlock here - if you have more pending messages than `InboxCapacity` you'll get stuck

Reply:
@zalmen currently just printing to log and ignoring the message. Let me know if you still want to cancel CP.